sirepo/pkcli/extract.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+u"""Operations run inside the report directory to extract data.
+
+:copyright: Copyright (c) 2019 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+
+from pykern import pkio
+from pykern import pkjson
+from sirepo import simulation_db
+import functools
+import sirepo
+import sys
+
+
+# These commands take json-encoded arguments, and then print their result as
+# json to stdout.
+def _extract_cmd(fn):
+    @functools.wraps(fn)
+    def wrapper(*args):
+        result = fn(*[pkjson.load_any(arg) for arg in args])
+        encoded = pkjson.dump_pretty(result)
+        sys.stdout.write(encoded)
+    return wrapper
+
+
+# These commands are always run with cwd set to the appropriate run_dir
+def _run_dir():
+    return pkio.py_path('.')
+
+
+def _params():
+    return pkjson.load_any(_run_dir().join('in.json'))
+
+
+@_extract_cmd
+def remove_last_frame():
+    template = sirepo.template.import_module(_params())
+    if hasattr(template, 'remove_last_frame'):
+        template.remove_last_frame(_run_dir())
+
+
+@_extract_cmd
+def background_percent_complete(is_running):
+    params = _params()
+    template = sirepo.template.import_module(params)
+    return template.background_percent_complete(
+        params.report, _run_dir(), is_running,
+    )
+
+
+# XX FIXME: can we avoid passing the whole data dict through here?
```

Comment:
The main issue is the `'models' in data` check, which IIUC is actually a proxy for checking whether this is the first time we're extracting the result from a given `run_dir`, because somehow the distributed state machine managed by javascript+flask makes that work as a way to check. I think this function really only needs a few bits of `data`, and if we sat down and traced it out then it would be pretty straightforward to decouple it from the js/flask details. I just didn't trust myself to untangle that at the same time I was waist-deep in implementing the extract job machinery.

Reply:
I think we'll have to wait on this optimization. See #1557 for discussion